src/main/java/com/google/sps/servlets/KeywordSearchServlet.java

```diff
@@ -0,0 +1,313 @@
+    List<String> list = new ArrayList<String>();
+    List<Map.Entry<String, Integer>> mapList =
+        new ArrayList<Map.Entry<String, Integer>>(map.entrySet());
+    Collections.sort(
+        mapList,
+        new Comparator<Map.Entry<String, Integer>>() {
+          public int compare(Map.Entry<String, Integer> o1, Map.Entry<String, Integer> o2) {
+            return (o2.getValue()).compareTo(o1.getValue());
+          }
+        });
+    for (int i = 0; i < mapList.size(); i++) {
+      list.add(mapList.get(i).getKey().toString());
+    }
+    return list;
+  }
+
+  /**
+   * Returns the keyword values for a map.
+   *
+   * @param map Map containing the keywords and their weighted occurrences
+   * @return List containing the keyword values only
+   */
+  public static List<Integer> getKeywordMapValues(Map<String, Integer> map) {
+    List<Integer> list = new ArrayList<Integer>();
+    List<Map.Entry<String, Integer>> mapList =
+        new ArrayList<Map.Entry<String, Integer>>(map.entrySet());
+    Collections.sort(
+        mapList,
+        new Comparator<Map.Entry<String, Integer>>() {
+          public int compare(Map.Entry<String, Integer> o1, Map.Entry<String, Integer> o2) {
+            return (o2.getValue()).compareTo(o1.getValue());
+          }
+        });
+    for (int i = 0; i < mapList.size(); i++) {
+      list.add(mapList.get(i).getValue());
+    }
+    return list;
+  }
+
+  /**
+   * Returns a list of words contained in a string.
+   *
+   * @param str String to be processed
+   * @return List containing every word in a string
+   */
+  public static List<String> getSeparateWords(String str) {
+    // TODO: handle single quoted strings (right now keeps them in for
+    // the common case of contractions
+
+    // Split the string based on a regex: splits at every space and at
+    // most common non-alphabet characters
+    String[] list = str.split("([\\s.,!?:;()\\[\\]&\"\\s+])");
+    // Trim all whitespace
+    for (int i = 0; i < list.length; i++) {
+      list[i] = list[i].replaceAll("\\s+", "");
+    }
+    List<String> listArr = new ArrayList<String>(Arrays.asList(list));
+    // Remove all empty strings
+    listArr.removeAll(Arrays.asList("", null));
+    return listArr;
+  }
+
+  /**
+   * Returns a map of the occurrences of words in a string.
+   *
+   * @param input String to be processed
+   * @return Map containing word keys and instance count values
+   */
+  public static Map<String, Integer> wordCount(String input) {
+    List<String> words = KeywordSearchServlet.getSeparateWords(input);
+    // process title words
+    Map<String, Integer> map = new HashMap<String, Integer>();
+    for (String w : words) {
+      w = w.toLowerCase();
+      if (map.containsKey(w)) {
+        map.put(w, map.get(w) + 1);
+      } else {
+        map.put(w, 1);
+      }
+    }
+    return map;
+  }
+
+  /**
+   * Returns a the total number of weighted occurrences of keywords in a list.
```

Comment:
The doc seems very ambiguous. Please give a formula or example. The param names could be improved as well.

Reply:
Got it!